addons/website/models/website_controller_page.py

```diff
@@ -0,0 +1,77 @@
+
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo.addons.http_routing.models.ir_http import slugify
+from odoo import api, fields, models
+
+
+class WebsiteControllerPage(models.Model):
+    _name = 'website.controller.page'
+    _inherits = {'ir.ui.view': 'view_id'}
+    _inherit = [
+        'website.published.multi.mixin',
+        'website.searchable.mixin',
+    ]
+    _description = 'Model Page'
+    _order = 'website_id'
+
+    view_id = fields.Many2one('ir.ui.view', string='View', required=True, ondelete="cascade")
+    menu_ids = fields.One2many('website.menu', 'controller_page_id', 'Related Menus')
+
+    website_id = fields.Many2one(related='view_id.website_id', store=True, readonly=False, ondelete='cascade')
+
+    # Bindings to model/records, to expose the page on the website.
+    # Route: /model/<string:page_name_slugified>
+    page_name = fields.Char(string="Name", help="The name is used to generate the URL and is shown in the browser title bar", required=True)
+    name_slugified = fields.Char(compute="_compute_name_slugified", store=True,
+        string="URL", help="The name of the page usable in a URL")
+    page_type = fields.Selection(selection=[("listing", "Listing"), ("single", "Single record")],
+        default="listing", string="Page Type", help="The type of the page. If set, it indicates whether the page displays a list of records or a single record")
+    record_domain = fields.Char(string="Domain", help="Domain to restrict records that can be viewed publicly")
+    default_layout = fields.Selection(
+        selection=[
+            ('grid', "Grid"),
+            ('list', "List"),
+        ],
+        default="grid",
+    )
+
+    def _get_default_layout(self, view_id):
+        return self.sudo().search_read([("view_id", "=", view_id)], ["default_layout"])[0]["default_layout"]
+
+    @api.constrains('view_id', 'model_id', "model")
+    def _check_user_has_model_access(self):
+        for record in self:
+            self.env[record.model_id.model].check_access_rights('read')
+
+    @api.model
+    def _get_page_url(self, page_name, additional_path=None):
+        name_slugified = slugify(page_name)
+        url = f"/model/{name_slugified}"
+        if additional_path:
+            additional_path = f"/{additional_path}" if not additional_path.startswith("/") else additional_path
```

Comment:
I don't think it's exploitable but constructing url by hand with user input is always risky, escape the additional_path to be sure

Reply:
and even better, this method does not need to be on the model, it's only called once in the controller. Can be moved there